### DISCUSSION TITLE
MySQL에서 대량의 데이터(500만개 이상)를 Insert해야하는 경우엔 어떻게 해야할까요?7/14

### DISCUSSION THREAD

**Discussion by SionBackEnd** [Team-Psyduck/tech-sharing#71] [General]
[오늘의 알고리즘 프로그래머스 주식가격](https://school.programmers.co.kr/learn/courses/30/lessons/42584)

**Comment by eatingbug** [comment]
- Bulk Insert
  - INSERT 문을 각 행에 별도로 실행하는 대신, 여러 행을 한번에 실행하는 Bulk Insert 를 사용한다.
  - 쿼리 비용이 줄어들고, 전체적인 성능이 향상된다.
- Transactions
  - SQL 트랜잭션을 사용하여 데이터를 묶어서 삽입할 수 있다. 데이터를 각각 삽입 후에 commit 을 수행하지 않도록 하여 성능을 향상시킨다.
- Disable Indexes
  - 삽입 전에 인덱스를 비활성화하면 성능이 향상될 수 있다. 
- Load Data INFILE
  - 매우 큰 데이터셋을 삽입할 경우 Load Data INFILE 을 사용하면 효과적이다. 텍스트 파일의 내용을 테이블에 직접 삽입하는 방식
- Tunning
  - MySQL 설정을 튜닝하여 성능을 향상시킬 수 있다.
  - `innodb_buffer_pool_size`, `max_allowed_packet` ...

## 참고자료
- [Blog - MySQL 대량 데이터 Insert 성능 향상](https://blog.naver.com/PostView.nhn?blogId=islove8587&logNo=222326263425)
- [Document - How to Quickly Insert Data Into MariaDB](https://mariadb.com/kb/en/how-to-quickly-insert-data-into-mariadb/)

**Comment by SionBackEnd** [comment]
삽입 방법으로는 4가지 정도 설명 가능합니다. 
1. Using Big Transactions
한 행에 많은 삽입을 수행할 때 모든 행에 대해 전체 트랜잭션을 수행하지 않게 하기 위해 이를 BEGIN / END로 래핑해서 사용합니다. 속도는 거의 1000배 정도 빨라질 수 있습니다. 
2. Multi-Value Inserts
다중 값 행 삽입을 사용하여 한 번에 여러 행을 삽입하는 방법입니다. 
3. Inserting Data into Serveral Tables at Once
한 번에 여러 테이블에 데이터를 삽입해야 하는 경우 가장 좋은 방법은 다중 행 문을 활성화하고 많은 삽입을 서버에 보내는 방법입니다. 
4. Loading Text File
LOAD DATA INFILE 방법이 가장 빠른 방법은 SQL 구문 분석과정이 없고, 데이터는 큰 블록에서 읽게됩니다. 작업 시작시 테이블이 비어 있으면 작업 중에 공유하지 않은 모든 인덱스가 비활성화 됩니다. 

[참고](https://blog.naver.com/PostView.nhn?blogId=islove8587&logNo=222326263425)
